docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,921 @@
+
+#### Machine Spec
+
+Machine's spec is going to be improved to allow 3rd party to extend the semantic of the new Machine's `Ready` condition
+as well to standardize the concept of Machine's `Availability`.
+
+Below you can find the relevant fields in Machine Status v1beta2, after v1beta1 removal (end state);
+After golang types, you can find a summary table that also shows how changes will be rolled out according to K8s deprecation rules.
+
+```go
+type MachineSpec struct {
+    
+    // MinReadySeconds is the minimum number of seconds for which a Node for a newly created machine should be ready before considering the replica available.
+    // Defaults to 0 (machine will be considered available as soon as the Node is ready)
+    // +optional
+    MinReadySeconds int32 `json:"minReadySeconds,omitempty"`
+
+    // If specified, all readiness gates will be evaluated for Machine readiness.
+    // A Machine is ready when `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are "True"; 
+    // if other conditions are defined this field, those conditions should be "True" as well for the Machine to be ready.
+    // +optional
+    // +listType=map
+    // +listMapKey=conditionType
+    ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`
+
+    // Other fields...
+}
+
+// MachineReadinessGate contains the reference to a Machine condition to be used as readiness gates.
+type MachineReadinessGate struct {
+    // ConditionType refers to a condition in the Machine's condition list with matching type.
+    // Note: Both  Cluster API conditions or conditions added by 3rd party controller can be used as readiness gates.
+    ConditionType string `json:"conditionType"`
+}
+```
+
+| v1beta1 (current)      | v1Beta2 (tentative Q1 2025) | v1beta1 removal (tentative Q1 2026) |
+|------------------------|-----------------------------|-------------------------------------|
+| `ReadinessGates` (new) | `ReadinessGates`            | `ReadinessGates`                    |
+| other fields...        | other fields...             | other fields...                     |
+
+Notes:
+- Both `MinReadySeconds` and `ReadinessGates` should be treated as other in-place propagated fields (changing this should not trigger rollouts).
+- Similarly to Pod's `ReadinessGates`, also Machine's `ReadinessGates` accept only conditions with positive polarity; 
+  The Cluster API project might revisit this in future to stay aligned with Kubernetes or if there are use cases justifying this change.
+
+#### Machine Print columns
+
+| Current           | To be                         |
+|-------------------|-------------------------------|
+| `NAME`            | `NAME`                        |
+| `CLUSTER`         | `CLUSTER`                     |
+| `NODE NAME`       | `PAUSED` (new) (*)            |
+| `PROVIDER ID`     | `NODE NAME`                   |
+| `PHASE` (deleted) | `PROVIDER ID`                 |
```

Comment:
Going to sound like a broken record, apologies :D ! One of the goals is to improve the user friendliness of internals. I _feel_ a bit though that the current `phase` field communicates better that an explosion of fields

Reply:
Let's chat about this, I'm sure we can find our way to what matters for users.

Reply:
+ 1 on "this is FANTASTIC and AMAZING work". Thanks @fabriziopandini.
also argee on that the status.ready(bool type) and status.phase(enum type) communicate better than a set of fields or conditions. Although if it can be calculated by the high level consumer or internal code of Cluster API, what if Cluster API object controller calculates it and record in status.ready and status.phase ?
btw @sbueringer i think status.phase is used to show the current provisioning or operation state, e.g. Creating/Upgrading/ScalingUp etc.

Reply:
> btw @sbueringer i think status.phase is used to show the current provisioning or operation state, e.g. Creating/Upgrading/ScalingUp etc.

Depends on the object. I was here referring to the Machine object (because this table here is about the Machine object). MachineDeployment phase has also ScalingUp & ScaldingDown.

Reply:
Resolving because I'm consolidating all the discussions about status.phases into https://github.com/kubernetes-sigs/cluster-api/pull/10897#discussion_r1685919394